### PR TITLE
Add environment variables for OTLP ingest

### PR DESCRIPTION
**Pull request by mx-psi**
### What does this PR do?

Add environment variables for most commonly used OTLP ingest features.

### Motivation

Provide a way to configure OTLP features through environment variables.

### Additional Notes

<details>
<summary>Script used to generate the code</summary>

```go
package main

import (
	"fmt"
	"reflect"
	"strings"

	"go.opentelemetry.io/collector/receiver/otlpreceiver"
)

// metricsConfig defines the metrics exporter specific configuration options
type MetricsConfig struct {
	// Quantiles states whether to report quantiles from summary metrics.
	// By default, the minimum, maximum and average are reported.
	Quantiles bool `mapstructure:"report_quantiles"`

	// SendMonotonic states whether to report cumulative monotonic metrics as counters
	// or gauges
	SendMonotonic bool `mapstructure:"send_monotonic_counter"`

	// DeltaTTL defines the time that previous points of a cumulative monotonic
	// metric are kept in memory to calculate deltas
	DeltaTTL int64 `mapstructure:"delta_ttl"`

	ExporterConfig metricsExporterConfig `mapstructure:",squash"`

	// TagCardinality is the level of granularity of tags to send for OTLP metrics.
	TagCardinality string `mapstructure:"tag_cardinality"`

	// HistConfig defines the export of OTLP Histograms.
	HistConfig histogramConfig `mapstructure:"histograms"`
}

// histogramConfig customizes export of OTLP Histograms.
type histogramConfig struct {
	// Mode for exporting histograms. Valid values are 'distributions', 'counters' or 'nobuckets'.
	//  - 'distributions' sends histograms as Datadog distributions (recommended).
	//  - 'counters' sends histograms as Datadog counts, one metric per bucket.
	//  - 'nobuckets' sends no bucket histogram metrics. .sum and .count metrics will still be sent
	//    if `send_count_sum_metrics` is enabled.
	//
	// The current default is 'distributions'.
	Mode string `mapstructure:"mode"`

	// SendCountSum states if the export should send .sum and .count metrics for histograms.
	// The current default is false.
	SendCountSum bool `mapstructure:"send_count_sum_metrics"`
}

// metricsExporterConfig provides options for a user to customize the behavior of the
// metrics exporter
type metricsExporterConfig struct {
	// ResourceAttributesAsTags, if set to true, will use the exporterhelper feature to transform all
	// resource attributes into metric labels, which are then converted into tags
	ResourceAttributesAsTags bool `mapstructure:"resource_attributes_as_tags"`

	// InstrumentationLibraryMetadataAsTags, if set to true, adds the name and version of the
	// instrumentation library that created a metric to the metric tags
	InstrumentationLibraryMetadataAsTags bool `mapstructure:"instrumentation_library_metadata_as_tags"`
}

type Config struct {
	Receiver otlpreceiver.Config `mapstructure:"receiver"`
	Metrics  MetricsConfig       `mapstructure:"metrics"`
}

func forEachField(prefix string, config interface{}, f func(yamlKey string, value interface{})) {
	v := reflect.ValueOf(config)
	t := reflect.TypeOf(config)

	visibleFields := reflect.VisibleFields(t)
	for _, field := range visibleFields {
		if !field.IsExported() {
			continue
		}

		key := strings.SplitN(field.Tag.Get("mapstructure"), ",", 2)[0]
		fieldVal := v.FieldByIndex(field.Index)
		yamlKey := prefix
		if key != "" {
			yamlKey = prefix + "." + key
		}

		if strings.HasPrefix(yamlKey, ".receiver.grpc") || strings.HasPrefix(yamlKey, ".receiver.http") {
			// HACK: Skip embedded fields, we only want the promoted fields which have 'protocols' in the path
			continue
		}

		if fieldVal.Kind() == reflect.Ptr {
			if fieldVal.Elem().Kind() == reflect.Struct {
				fieldVal = fieldVal.Elem()
			} else {
				continue
			}
		}

		if fieldVal.Kind() == reflect.Struct {
			forEachField(yamlKey, fieldVal.Interface(), f)
		} else {
			f(yamlKey, fieldVal.Interface())
		}
	}
}

func printConfig(yamlKey string, _ interface{}) {
	fmt.Printf("\tconfig.BindEnv(OTLPSection + %q)\n", yamlKey)
}

func main() {
	receiver := otlpreceiver.NewFactory()
	otlpconfig := *(receiver.CreateDefaultConfig().(*otlpreceiver.Config))
	config := Config{
		Receiver: otlpconfig,
		Metrics: MetricsConfig{
			SendMonotonic: true,
			DeltaTTL:      3600,
			Quantiles:     true,
			ExporterConfig: metricsExporterConfig{
				ResourceAttributesAsTags:             false,
				InstrumentationLibraryMetadataAsTags: false,
			},
			TagCardinality: "low",
			HistConfig: histogramConfig{
				Mode:         "distributions",
				SendCountSum: false,
			},
		},
	}

	forEachField("", config, printConfig)
}
```

</details>

Some fields related to TLS are not exposed as environment variables. Their support on the Agent needs to be considered more closely because of the inter-Agent communication, so I think it's okay if they are not added for now.

### Possible Drawbacks / Trade-offs

- The Viper setup is fragile, so care needs to be taken that the default values are not overridden with the zero value.
- Environment variable names are a bit too long (e.g. `DD_OTLP_CONFIG_METRICS_INSTRUMENTATION_LIBRARY_METADATA_AS_TAGS`), but they are predictable. We can add shorter names for commonly used features if needed in the future. 
- You can't set an empty section as in
   ```yaml
   otlp_config:
     receiver:
        protocols:
          grpc:
   ```
    through environment variables. This is still an open problem.

### Describe how to test/QA your changes

Set the OTLP ingest feature through environment variables (e.g. set `DD_OTLP_CONFIG_RECEIVER_PROTOCOL_GRPC_ENDPOINT` to `0.0.0.0:4317`)

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/otlp/config.go [hunk 1]
```diff
@@ -29,16 +30,34 @@ func portToUint(v int) (port uint, err error) {
...
+	// If we only had the stuff below, the nil sections would be ignored.
+	// We want to take into account nil-but-present sections.
+	cfgMap := colConfig.NewMapFromStringMap(cfg.GetStringMap(section))
+
+	// Step two works around https://github.com/spf13/viper/issues/1012
+	// we check every key manually, and if it belongs to the OTLP receiver section,
+	// we set it. We need to do this to account for environment variable values.
+	prefix := section + "."
+	for _, key := range cfg.AllKeys() {
+		if strings.HasPrefix(key, prefix) && cfg.IsSet(key) {
```
Comment (mx-psi) [inline pkg/otlp/config.go:49]: To make things more fun, `AllKeys` returns all known keys, even those that are not set, so I have to check if they are set before setting them in the config map.

File: pkg/config/otlp.go [hunk 2]
```diff
@@ -87,3 +89,36 @@ func promoteExperimentalOTLP(cfg Config) {
...
 }
+
+// setupEnvironmentVariables sets up the environment variables associated with different OTLP ingest settings:
+// If there are changes in the OTLP receiver configuration, they should be reflected here.
+//
+// We don't need to set the default value: it is dealt with at the unmarshaling level
+// since we get the configuration through GetStringMap
+//
+// We are missing TLS settings: since some of them need more work to work right they are not included here.
+func setupEnvironmentVariables(config Config) {
```
Comment (vickenty) [inline pkg/config/otlp.go:100]: Should probably have `otlp` in the name somewhere, because the func is in the common `config` package.

**Review by vickenty — CHANGES_REQUESTED** [review]

File: pkg/config/otlp.go (hunk 2 above)
Comment (mx-psi) [reply pkg/config/otlp.go]: done!

**Review by vickenty — APPROVED** [review]

**Review by maycmlee — APPROVED** [review]
👍  for docs.

**Review by KSerrania — APPROVED** [review]

**Pull request merged**